unit-tests/src/test/scala/java/net/UdpSocketTest.scala

```diff
@@ -0,0 +1,139 @@
+      assertNotEquals("getsockname", -1, gsnStatus)
+
+      // Now use port.
+      val outSocket = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)
+      assertNotEquals("socket", -1, outSocket)
+
+      try {
+        val outAddr      = alloc[sockaddr]
+        val outAddrInPtr = outAddr.asInstanceOf[Ptr[sockaddr_in]]
+        outAddrInPtr.sin_family = AF_INET.toUShort
+        outAddrInPtr.sin_addr.s_addr = localhostInetAddr
+        outAddrInPtr.sin_port =
+          inAddrInfo.asInstanceOf[Ptr[sockaddr_in]].sin_port
+
+        val outData =
+          """
+          |Four Freedoms -
+          |   Freedom of speech
+          |   Freedom of worship
+          |   Freedom from want
+          |   Freedom from fear
+          """.stripMargin
+
+        val nBytesSent = sendto(outSocket,
+                                toCString(outData),
+                                outData.length.toULong,
+                                0,
+                                outAddr,
+                                sizeof[sockaddr].toUInt)
+        assertEquals("sendto", outData.size, nBytesSent)
+
+        // There is a "pick your poison" design choice here.
+        // inSocket is set O_NONBLOCK to eliminate the possibility
+        // that a bad sendto() or readfrom() implemenation would hang
+        // for a long time.
+        //
+        // This introduces the theoretical possiblity the sendto() above
+        // does not complete before recvfrom() looks for data, causing
+        // failure. Since this is send/recv pair is explicitly loopback,
+        // that is highly unlikely.
+        //
+        // If this race condition becomes bothersome, a Thead.sleep() can
+        // be inserted here.
+
+        /// Two tests using one inbound packet, save test duplication.
+
+        // Provide extra room to allow detecting extra junk being sent.
+        val maxInData = 2 * outData.length
+        val inData    = alloc[Byte](maxInData)
+
+        // Test not fetching remote address. Exercise last two arguments.
+        val nBytesPeekedAt =
+          recvfrom(inSocket,
+                   inData,
+                   maxInData.toUInt,
+                   MSG_PEEK,
+                   null.asInstanceOf[Ptr[sockaddr]],
+                   null.asInstanceOf[Ptr[socklen_t]])
+
+        // Friendlier code here and after the next recvfrom() would loop
+        // on partial reads rather than fail.
+        // Punt to a future generation. Since this is loopback and
+        // writes are small, if any bytes are ready, all should be.
+        assertEquals("recvfrom length", nBytesSent, nBytesPeekedAt)
+
+        // Test retrieving remote address.
+        val srcAddr    = alloc[sockaddr]
+        val srcAddrLen = alloc[socklen_t]
+        !srcAddrLen = sizeof[sockaddr].toUInt
+        val nBytesRecvd =
+          recvfrom(inSocket, inData, maxInData.toUInt, 0, srcAddr, srcAddrLen)
+
+        assertEquals("recvfrom length", nBytesSent, nBytesRecvd)
```

Comment:
This probably should have another helper message since it was already used in previous `assertEquals`.

Reply:
Good catch on same message in more than one place (for each of two messages).
I absolutely __detest__ such when I encounter them and am embarrassed to have
unintentionally created such myself.

Corrected for each of two messages.